### PR TITLE
Enable many photos. Fix bug with face outside image. Improve layout

### PR DESCRIPTION
**Pull request by DominikKwiatkowski**

### CODE REVIEW

File: app/src/main/java/com/libs/benchmark/BenchmarkLayout.java [hunk 1]
```diff
@@ -4,36 +4,53 @@
...
-
+        // TODO Clear models before every start!!!
         // Read supported models.
         for (String model : caller.getResources().getStringArray(R.array.models)) {
             if (userSettings.getBoolean(model + caller.getString(R.string.settings_benchModel_suffix), true)) {
                 supportedModels.add(new Pair<>(model, model + caller.getString(R.string.BenchmarkMode_ModelDatabaseName_Suffix)));
             }
         }
+        // Check if there is a supported model
+        if(supportedModels.size()==0){
```
Comment (Qubaef) [inline app/src/main/java/com/libs/benchmark/BenchmarkLayout.java:42]: Reformat the file.

File: app/src/main/java/com/libs/benchmark/BenchmarkLayout.java [hunk 2]
```diff
@@ -4,36 +4,53 @@
...
+            toastWrapper.showToast("Lacking supported model!!!", Toast.LENGTH_LONG);
+            caller.finish();
+
```
Comment (Qubaef) [inline app/src/main/java/com/libs/benchmark/BenchmarkLayout.java:45]: Unnecessary newline

File: app/src/main/java/com/libs/benchmark/BenchmarkLayout.java [hunk 3]
```diff
@@ -4,36 +4,53 @@
...
+        // TODO Clear models before every start!!!
         // Read supported models.
         for (String model : caller.getResources().getStringArray(R.array.models)) {
             if (userSettings.getBoolean(model + caller.getString(R.string.settings_benchModel_suffix), true)) {
                 supportedModels.add(new Pair<>(model, model + caller.getString(R.string.BenchmarkMode_ModelDatabaseName_Suffix)));
             }
         }
+        // Check if there is a supported model
+        if(supportedModels.size()==0){
+            toastWrapper.showToast("Lacking supported model!!!", Toast.LENGTH_LONG);
```
Comment (Qubaef) [inline app/src/main/java/com/libs/benchmark/BenchmarkLayout.java:43]: Do not use exclamation marks in UI

File: app/src/main/java/com/libs/benchmark/BenchmarkLayout.java [hunk 4]
```diff
@@ -4,36 +4,53 @@
...
 public class BenchmarkLayout implements LayoutClassInterface {
     private final ArrayList<Pair<String, String>> supportedModels = new ArrayList<>();
     private final AppCompatActivity caller;
 
-    private final LayoutClassInterface addPhotoLayout;
+    private final AddPhotoLayout addPhotoLayout;
     private final LayoutClassInterface displayResultLayout;
+    // This field is only to get data about state of database
+    private final ModelObject sampleModelObject;
+    ToastWrapper toastWrapper;
```
Comment (Qubaef) [inline app/src/main/java/com/libs/benchmark/BenchmarkLayout.java:28]: No access specifier

File: app/src/main/java/com/libs/benchmark/BenchmarkLayout.java [hunk 5]
```diff
@@ -4,36 +4,53 @@
...
 import com.R;
 import com.activities.AddFaceActivity;
+import com.common.ToastWrapper;
 import com.libs.globaldata.GlobalData;
+import com.libs.globaldata.ModelObject;
 
 import java.util.ArrayList;
 
 public class BenchmarkLayout implements LayoutClassInterface {
     private final ArrayList<Pair<String, String>> supportedModels = new ArrayList<>();
```
Comment (Qubaef) [inline app/src/main/java/com/libs/benchmark/BenchmarkLayout.java:21]: Change to List<>

File: app/src/main/java/com/libs/facerecognition/FacePreprocessor.java [hunk 6]
```diff
@@ -170,7 +170,16 @@ private Bitmap rotateAndTrimFace(Bitmap image, Face face) {
         List<Bitmap> cutFaces = new ArrayList<>();
 
         for (Face face : listOfFaces) {
-            cutFaces.add(rotateAndTrimFace(frame, face));
+            // Check if face is on image.
+            if(face.getBoundingBox().top<=FaceMargin ||
```
Comment (Qubaef) [inline app/src/main/java/com/libs/facerecognition/FacePreprocessor.java:174]: This should be excluded to separate private function with single responsibility

File: app/src/main/java/com/libs/globaldata/userdatabase/UserDatabase.java [hunk 7]
```diff
@@ -196,6 +199,14 @@ public UserRecord getUserRecord(String userName) {
...
     }
 
+    /**
+     * Get the number of currently added Users.
+     * @return int Number of users
+     */
+    public int getNumberOfUsers()
+    {
+        return getUsersArray().length;
+    }
```
Comment (Qubaef) [inline app/src/main/java/com/libs/globaldata/userdatabase/UserDatabase.java:209]: Missing newline

File: app/src/main/res/layout/activity_benchmark_add_photo.xml [hunk 8]
```diff
@@ -16,10 +16,10 @@
         android:layout_marginTop="156dp"
         android:layout_marginEnd="4dp"
         android:background="#00000000"
-        app:icon="@drawable/right"
+        app:icon="@drawable/right_array"
```
Comment (Qubaef) [inline app/src/main/res/layout/activity_benchmark_add_photo.xml:19]: Why are these arrows called "array"? Whats the connection with the array? Why are you assuming that they are used to iterate over array? Change it to "ic_right_arrow"

File: app/src/main/res/layout/activity_benchmark_add_photo.xml [hunk 9]
```diff
@@ -16,10 +16,10 @@
...
         app:iconTint="@color/black"
         app:layout_constraintEnd_toEndOf="parent"
-        app:layout_constraintHorizontal_bias="0.438"
+        app:layout_constraintHorizontal_bias="0.41"
```
Comment (Qubaef) [inline app/src/main/res/layout/activity_benchmark_add_photo.xml:22]: Why are we relying on some random value? Change to dp, or make relative to other components.

File: app/src/main/res/layout/activity_benchmark_add_photo.xml [hunk 10]
```diff
@@ -84,10 +84,10 @@
         android:layout_marginTop="156dp"
         android:layout_marginEnd="1dp"
         android:background="#00000000"
-        app:icon="@drawable/left"
+        app:icon="@drawable/left_array"
         app:iconTint="@color/black"
         app:layout_constraintEnd_toStartOf="@+id/benchSelectedImage"
-        app:layout_constraintHorizontal_bias="0.4"
+        app:layout_constraintHorizontal_bias="0.529"
```
Comment (Qubaef) [inline app/src/main/res/layout/activity_benchmark_add_photo.xml:90]: Why are we relying on some random value? Change to dp, or make relative to other components.

File: app/src/main/res/layout/activity_benchmark_mode.xml [hunk 11]
```diff
@@ -25,10 +25,11 @@
       android:layout_width="wrap_content"
       android:layout_height="wrap_content"
       android:layout_marginStart="104dp"
-      android:layout_marginTop="104dp"
+      android:layout_marginTop="220dp"
       android:layout_marginEnd="104dp"
       android:text="@string/BenchmarkMode_MakePhoto_Text"
       app:layout_constraintEnd_toEndOf="parent"
+      app:layout_constraintHorizontal_bias="0.0"
```
Comment (Qubaef) [inline app/src/main/res/layout/activity_benchmark_mode.xml:32]: Change to 0dp just to make things unified

File: app/src/main/res/layout/activity_benchmark_mode.xml [hunk 12]
```diff
@@ -43,4 +44,30 @@
...
       app:layout_constraintStart_toStartOf="parent" />
+
+  <TextView
+      android:id="@+id/numberOfFacesView"
+      android:layout_width="wrap_content"
+      android:layout_height="wrap_content"
+      android:layout_marginStart="180dp"
+      android:layout_marginEnd="180dp"
+      android:layout_marginBottom="18dp"
+      android:text="TextView"
```
Comment (Qubaef) [inline app/src/main/res/layout/activity_benchmark_mode.xml:55]: No need to add text here. If you want it to be displayed in preview window, change to someting more self-explanatory like a name of the field.

File: app/src/main/java/com/libs/facerecognition/FacePreprocessor.java [hunk 13]
```diff
@@ -26,7 +26,7 @@
 public class FacePreprocessor {
     public final String Tag = "ImagePreProcessor";
     private final FaceDetector faceDetector;
-
+    private final int FaceMargin = 10;
```
Comment (Qubaef) [inline app/src/main/java/com/libs/facerecognition/FacePreprocessor.java:29]: I tested this locally and I fell like this margin is way to high. Couldn't it be 0?

File: app/src/main/res/layout/activity_benchmark_add_photo.xml (hunk 9 above)
Comment (DominikKwiatkowski) [reply app/src/main/res/layout/activity_benchmark_add_photo.xml:22]: https://www.google.com/search?q=what+is+bias+in+constraint+layout+android&oq=what+is+constraint+bias&aqs=chrome.1.69i57j0i22i30l2.6492j0j7&sourceid=chrome&ie=UTF-8

File: app/src/main/res/layout/activity_benchmark_add_photo.xml (hunk 10 above)
Comment (DominikKwiatkowski) [reply app/src/main/res/layout/activity_benchmark_add_photo.xml:90]: https://www.google.com/search?q=what+is+bias+in+constraint+layout+android&oq=what+is+constraint+bias&aqs=chrome.1.69i57j0i22i30l2.6492j0j7&sourceid=chrome&ie=UTF-8

File: app/src/main/res/layout/activity_benchmark_mode.xml (hunk 11 above)
Comment (DominikKwiatkowski) [reply app/src/main/res/layout/activity_benchmark_mode.xml]: +1

File: app/src/main/java/com/libs/facerecognition/FacePreprocessor.java (hunk 13 above)
Comment (DominikKwiatkowski) [reply app/src/main/java/com/libs/facerecognition/FacePreprocessor.java]: +1

File: app/src/main/res/layout/activity_benchmark_mode.xml (hunk 12 above)
Comment (DominikKwiatkowski) [reply app/src/main/res/layout/activity_benchmark_mode.xml]: +1

File: app/src/main/java/com/libs/benchmark/BenchmarkLayout.java (hunk 5 above)
Comment (DominikKwiatkowski) [reply app/src/main/java/com/libs/benchmark/BenchmarkLayout.java]: +1

File: app/src/main/java/com/libs/facerecognition/FacePreprocessor.java [hunk 14]
```diff
@@ -170,10 +170,22 @@ private Bitmap rotateAndTrimFace(Bitmap image, Face face) {
...
+            } else {
+                cutFaces.add(rotateAndTrimFace(frame, face));
+            }
         }
 
         Log.d(Tag, "Preprocessed all images");
         return cutFaces;
     }
+
+    private boolean isFaceOnImage(Face face, Bitmap image) {
```
Comment (Qubaef) [inline app/src/main/java/com/libs/facerecognition/FacePreprocessor.java:185]: Missing javadoc

File: app/src/main/java/com/libs/globaldata/userdatabase/UserDatabase.java (hunk 7 above)
Comment (DominikKwiatkowski) [reply app/src/main/java/com/libs/globaldata/userdatabase/UserDatabase.java:209]: +1

**Review by Qubaef — APPROVED** [review]

**Pull request merged**